File: leetcode/1-100/53-maximum-subarray/bench/maxsub_bench.py
"""Bench mirror of maxsub_bench.kara — Kadane over a batch of LCG-generated arrays,
i64 sink of per-array answers. CPython. See ../README.md § Benchmarks.

CPython is ~1-2 orders of magnitude slower on this tight scalar loop, so it is
excluded from the headline table by default (KARA_BENCH_INCLUDE_PY=1 to include).
"""


def main() -> None:
    m = 1103515245        # glibc LCG multiplier
    inc = 12345           # glibc LCG increment
    modulus = 2147483648  # 2^31
    windows = 120000      # number of simulated input arrays
    w = 1000              # length of each array

    state = 1             # LCG seed
    sink = 0
    for _ in range(windows):
        state = (state * m + inc) % modulus
        v0 = (state % 100) - 50
        best = here = v0
        for _ in range(1, w):
            state = (state * m + inc) % modulus
            v = (state % 100) - 50
            here = here + v if here + v > v else v
            if here > best:
                best = here
        sink += best
    print(sink)


if __name__ == "__main__":
    main()

File: leetcode/1-100/53-maximum-subarray/bench/maxsub_bench.rs
// Bench mirror of maxsub_bench.kara — Kadane over a batch of LCG-generated arrays,
// i64 sink of per-array answers. rustc -O. See ../README.md § Benchmarks.

fn main() {
    let m: i64 = 1103515245; // glibc LCG multiplier
    let inc: i64 = 12345; // glibc LCG increment
    let modulus: i64 = 2147483648; // 2^31
    let windows: i64 = 120000; // number of simulated input arrays
    let w: i64 = 1000; // length of each array

    let mut state: i64 = 1; // LCG seed
    let mut sink: i64 = 0;
    let mut k: i64 = 0;
    while k < windows {
        state = (state * m + inc) % modulus;
        let v0 = (state % 100) - 50;
        let mut best = v0;
        let mut here = v0;
        let mut j = 1i64;
        while j < w {
            state = (state * m + inc) % modulus;
            let v = (state % 100) - 50;
            here = if here + v > v { here + v } else { v };
            if here > best {
                best = here;
            }
            j += 1;
        }
        sink += best;
        k += 1;
    }
    println!("{}", sink);
}
